Fix device propagation for checkpointing

dvrogozh opened the pull request:
Fixes: #128478

In backward() implementation checkpointing code was quering device type from the rng_state tensors saved on forward(). These tensors are CPU only tensors and don't carry device information with them. As a result CUDA device was assumed as a default. Which is not correct if user runs on some other device. For example, on XPU.

This patch saves full device information on forward() and uses it on backward() to get device type. Previously forward save only device index.

cc @gujinghui @EikanWang @fengyuan14 @guangyey

dvrogozh commented:
Some tests are failing:
```
  File "/var/lib/jenkins/workspace/test/test_utils.py", line 518, in test_infer_device_state_recursive_meta
    device_type = _infer_device_type(inp)
```

I forgot to check whether _infer_device_type() prototype of which I have changed is called from anywhere else. I assumed that's internal function. Will move PR to draft for the fix.

dvrogozh commented on torch/utils/checkpoint.py:
```diff
@@ -126,30 +126,33 @@ def get_device_type() -> str:
-    device_types_set = set(device_types)
+    return _infer_device_type_from_devices(devices)
+
+def _infer_device_type_from_devices(devices):
```
I initially thought to rename above one as `_infer_device_type_from_tensors`, but it occurs it's used in some tests. So, to minimize the change I left it intact.

dvrogozh commented:
> I forgot to check whether _infer_device_type() prototype of which I have changed is called from anywhere else.

Fixed. Ready for review.

dvrogozh commented:
@albanD @soulitz @fmassa : can you, please, help review?

EikanWang commented:
@gujinghui , has ipex encountered the issue?

EikanWang commented on torch/utils/checkpoint.py:
```diff
@@ -159,31 +162,31 @@ def add_device_types(arg):
 
 def set_device_states(devices, states) -> None:
-    device_module = _get_device_module(_infer_device_type(*states))
+    device_module = _get_device_module(_infer_device_type_from_devices(devices))
```
`_get_device_module` supports different devices and `_infer_device_type` also supports different devices. May I know why we need to such changes?

dvrogozh replied:
`_get_device_module` returns appropriate module (`torch.cuda` or `torch.xpu`) depending on the given device type.

`_infer_device_type` returns single device type on which given tensors are allocated (these are rng_state tensors). If tensors are on different devices, one will be returned with the warning printed. See logic in the function.

The whole problem and a bug is that rng_state tensors are CPU tensors and does not carry device information with them at all as far as I can tell. So, `_infer_device_type` will  always return `cuda` device type since it's the default for the case when all tensors are cpu ones (again - see logic in the function).

My change is to deduce device type not from the rng_state tensors, but from devices list saved on `forward()`. Previously this information was just discarded and only device index was saved.

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -159,31 +162,31 @@ def add_device_types(arg):
+    return fwd_devices, fwd_device_states
 
 
 def set_device_states(devices, states) -> None:
```
Could we have set_device_states accept a "device" argument?
Then for both callsites (`use_reentrant={True,False}`) it is easy to access the device from backward, for `True` (custom autograd Function) path, you can use `ctx.device`, and for the `False` (saved tensor hook) path, you can use the closed over "device" variable.

dvrogozh replied:
Yes. Done.

soulitzer reviewed (commented):
Nice catch! But perhaps the support can be added in a different way? (~4 LoC total)

gujinghui commented:
> @gujinghui , has ipex encountered the issue?

Did not see this before...

dvrogozh committed:
Fix device propagation for checkpointing

Fixes: #128478

In backward() device type was queried from the rng_state tensors
saved on forward(). These tensors are CPU only tensors and don't
carry device information with them. As a result CUDA device was
assumed as a default. Which is not correct if user runs on some
other device. For example, on XPU.

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh commented:
Can someone, please, approve ci run?

dvrogozh commented:
> Could we have set_device_states accept a "device" argument?

@soulitzer : I've changed `set_device_states` to accept `device` argument. However, this function is being exported from the module, so BC Linter complains. It's not formally documented however and I don't see it is being used elsewhere in pytorch project. Can you, please, advice on the next step?
```
$ fgrep -rsn set_device_states
docs/source/conf.py:1868:    "set_device_states",
torch/utils/checkpoint.py:37:    "set_device_states",
torch/utils/checkpoint.py:168:# To consider:  maybe get_device_states and set_device_states should reside in torch/random.py?
torch/utils/checkpoint.py:190:def set_device_states(device_type, devices, states) -> None:
torch/utils/checkpoint.py:289:                    set_device_states(ctx.device, ctx.fwd_devices, ctx.fwd_device_states)
torch/utils/checkpoint.py:1394:                    set_device_states(device, fwd_devices, fwd_device_states)
```

soulitzer commented:
Doing a quick github search, I see it being used in some libraries that implement checkpoint-y things. To avoid bc-breaking, let's add it as an optional kwarg instead and add a comment that this arg is necessary if you are dealing with non-cuda devices.

dvrogozh commented:
> To avoid bc-breaking, let's add it as an optional kwarg instead and add a comment that this arg is necessary if you are dealing with non-cuda devices.

@soulitzer : Done. Please, take a look. If looks good, please, help approve ci run.

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -187,8 +187,17 @@ def add_device_ids(arg):
 
-def set_device_states(devices, states) -> None:
-    device_module = _get_device_module(_infer_device_type(*states))
+def set_device_states(devices, states, **kwargs) -> None:
```
Maybe instead `def set_device_states(devices, states, *, device_type="cuda")`?

dvrogozh replied:
Changed

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -286,13 +295,13 @@ def backward(ctx, *args):
                 torch.set_rng_state(ctx.fwd_cpu_state)
                 if ctx.had_device_in_fwd:
-                    set_device_states(ctx.fwd_devices, ctx.fwd_device_states)
+                    set_device_states(ctx.fwd_devices, ctx.fwd_device_states, device_type=ctx.device)
```
This is a preexisting issue, but do you mind renaming `ctx.device` and `device` to be `device_type`. It confused me when was reading this.

dvrogozh replied:
Will do.

dvrogozh replied:
renamed `device` to `device_type`

dvrogozh committed:
Rename device to device_type in checkpoint.py

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -187,18 +187,26 @@ def add_device_ids(arg):
+        devices: Device ids to set states for.
+        states: States to set.
+        device_type: ``device_type`` of the devices to set states for. This is
+            required in case of non-cuda devices. Defailt is ``cuda``.
```
```suggestion
            required in case of non-cuda devices. Default is ``cuda``.
```

dvrogozh replied:
fixed

dvrogozh commented:
@soulitzer : I addressed a typo. Please, help review again. And another round of ci run is needed.

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -187,18 +187,26 @@ def add_device_ids(arg):
+        device_type: ``device_type`` of the devices to set states for. This is
+            required in case of non-cuda devices. Default is ``cuda``.
+    """
+    device_module = _get_device_module(device_type)
```
Sorry I missed this earlier, but to preserve existing behavior, we should have the arg default to None, and take the value of  `DefaultDeviceType.get_device_type()` instead.

dvrogozh replied:
> take the value of DefaultDeviceType.get_device_type() instead.

Done.

dvrogozh commented:
@soulitzer : changed default device type to the value of `DefaultDeviceType.get_device_type()` and modified `set_device_states` argument description accordingly.

dvrogozh commented:
@soulitzer : ci has passed. Can this PR be merged? Any other reviews needed from anyone?

soulitzer commented:
Thanks for the updates so far, it is looking great.

Some comments on the testing side of things:
- Is it tested anywhere CUDA continues to works as expected. It would be good to make sure this is captured somewhere.
- If possible, add a test for non-cuda device.

dvrogozh commented:
> Is it tested anywhere CUDA continues to works as expected. It would be good to make sure this is captured somewhere.

As far as I see there are a number of tests in the pytorch test suite dedicated to checkpointing, including cuda dedicated tests. ci for this PR has passed, so I hope that's covered. However, I am not familiar enough with pytorch ci to say whether checkpointing tests for cuda were executed or not. Can someone more familiar with ci clarify that? From my side, I can do manual try out on cuda with same huggingface use case where I initially found the issue (note though, I can do that only after 25th of June after my vacation).

https://github.com/pytorch/pytorch/blob/832fc3521150112c6b9b411e00664dc1f2104560/test/test_cuda.py#L2002
https://github.com/pytorch/pytorch/blob/ddb95dbb0d81fc73ec743b689c049a5d8b85ae44/test/test_functionalization_of_rng_ops.py#L288
https://github.com/pytorch/pytorch/blob/734891ac22a47a6bceeb4f89bb53f9727c80c920/test/test_utils.py#L194

> If possible, add a test for non-cuda device.

Well, such a test would be quite handy to make sure this won't regress in the future. But we need non-cuda device in ci environment for that. XPU might provide the one, but as far as I know it's not yet enabled for ci (@EikanWang?). I also not sure where such test should be created since what I see XPU has some tests forked from pytorch in other project here: https://github.com/intel/torch-xpu-ops/tree/main/test/xpu. I don't know if any other backend can provide background for the test plus I don't have locally environment other than cuda and xpu.

Anyway, from my side I can work on the XPU based test after June 25th right after my vacation. I would prefer to add something to https://github.com/pytorch/pytorch/blob/main/test/test_xpu.py.

dvrogozh committed:
xpu: add test_autocast_checkpointing

Adding test_autocast_checkpointing similar to existing cuda test.

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh commented:
@soulitzer : I've added `test_autocast_checkpointing` to test/test_xpy.py, it works after applying this PR and fails without. That's a test similar to the cuda test in test/test_cuda.py:
https://github.com/pytorch/pytorch/blob/dadc0ed4c8ea0cd46402d7ff39415b3464b121cd/test/test_cuda.py#L2002

I've also found in the logs of prev. run that cuda's test is getting executed within at least [linux-focal-cuda12.1-py3.10-gcc9 / test (default, 1, 5, linux.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/9555187898/job/26340192001?pr=128671#logs):
```
2024-06-17T23:49:40.3879411Z Executing ['/opt/conda/envs/py_3.10/bin/python', '-bb', 'test_cuda.py', '-m', 'not serial', '--shard-id=1', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests'] ... [2024-06-17 23:49:40.387437]
2024-06-17T23:51:28.9149824Z 
2024-06-17T23:51:28.9154681Z test_cuda 1/1 was successful, full logs can be found in artifacts with path test/test-reports/test_cuda_1.1_17ecef8793471569_.log
2024-06-17T23:51:28.9230856Z Running 150 items in this shard: <...> test/test_cuda.py::TestCuda::test_autocast_checkpointing <...>
```
I hope this clarifies whether that's tested and still works for cuda. @soulitzer, please, let me know if you anticipate any other tests added or any other experiments for cuda done.

soulitzer commented on test/test_xpu.py:
```diff
@@ -438,6 +439,19 @@ def test_autocast_torch_expect_builtin_promote(self):
+            (8, 8), device="xpu", dtype=torch.float16, requires_grad=True
+        )
+        with torch.autocast("xpu"):
+            output = checkpoint_sequential(model, 2, input, use_reentrant=True)
```
nit: Could you test with both use_reentrant=True and False to make sure both code paths are covered?

dvrogozh replied:
Done. Testing now with both `use_reentrant=True` and `use_reentrant=False`

soulitzer reviewed (approved):
Thanks, minor nit, looks great otherwise!

dvrogozh committed:
cuda: add reentrant=False case for test_autocast_checkpointing

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh commented:
> Thanks, minor nit, looks great otherwise!

@soulitzer : Addressed. Covering `use_reentrant=True` and `=False` now. Note, I extended cuda's test as well to align the coverage across backends. Also note, I did verify xpu test, but did not verify cuda's test locally since I don't have handy nvidia gpu setup at the moment - please, trigger ci test rerun to get this covered.

dvrogozh commented:
@dvrogozh : ci has finished. There is one failure which seems to me as unrelated to this PR. Is this undetected flakiness in ci? Also, I have manually tried modified cuda test on nvidia card and ran initial huggingface case where I found the issue with checkpointing - both work. I hope this PR can be merged now.

```
2024-06-25T19:51:58.8054276Z _ TestSDPACudaOnlyCUDA.test_flash_attention_vs_math_ref_grads_batch_size_8_seq_len_q_1024_seq_len_k_1024_head_dim_128_is_causal_True_dropout_p_0_22_bfloat16_scale_l1_cuda_bfloat16 _
2024-06-25T19:51:58.8054418Z Traceback (most recent call last):
2024-06-25T19:51:58.8055078Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2874, in setUp
2024-06-25T19:51:58.8055200Z     set_rng_seed(SEED)
2024-06-25T19:51:58.8055881Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1969, in set_rng_seed
2024-06-25T19:51:58.8056016Z     torch.manual_seed(seed)
2024-06-25T19:51:58.8056517Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_compile.py", line 31, in inner
2024-06-25T19:51:58.8056661Z     return disable_fn(*args, **kwargs)
2024-06-25T19:51:58.8057220Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 601, in _fn
2024-06-25T19:51:58.8057348Z     return fn(*args, **kwargs)
2024-06-25T19:51:58.8057871Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/random.py", line 46, in manual_seed
2024-06-25T19:51:58.8058017Z     torch.cuda.manual_seed_all(seed)
2024-06-25T19:51:58.8058601Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/cuda/random.py", line 127, in manual_seed_all
2024-06-25T19:51:58.8058740Z     _lazy_call(cb, seed_all=True)
2024-06-25T19:51:58.8059298Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/cuda/__init__.py", line 244, in _lazy_call
2024-06-25T19:51:58.8059409Z     callable()
2024-06-25T19:51:58.8059917Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/cuda/random.py", line 125, in cb
2024-06-25T19:51:58.8060068Z     default_generator.manual_seed(seed)
2024-06-25T19:51:58.8060276Z RuntimeError: CUDA error: unspecified launch failure
2024-06-25T19:51:58.8060817Z CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
```

dvrogozh commented:
@soulitzer : what would be next steps here? can PR be merged now?

soulitzer commented:
Sorry for the delay, yes this can be merged now! The failure looks unrelated.

soulitzer commented:
@pytorchbot merge

dvrogozh commented:
@soulitzer : some ci checks have failed on the merge. Hope these are unrelated to this PR. Should PR be merged manually?

```
test_mps.py::TestNNMPS::test_invalid_conv1d Fatal Python error: Segmentation fault
...
AssertionError: "Cannot\ copy\ out\ of\ meta\ tensor;\ no\ data!\ Please\ use\ torch\.nn\.Module\.to_empty\(\)\ instead\ of\ torch\.nn\.Module\.to\(\)\ when\ moving\ module\ from\ meta\ to\ a\ different\ device\." does not match "Cannot copy out of meta tensor; no data!
...
FAILED [1.9972s] test_cuda_expandable_segments.py::TestCudaMallocAsync::test_allocator_fuzz - torch.OutOfMemoryError: CUDA out of memory.
```

soulitzer commented:
@pytorchbot merge -i
